apps/api/migrations/preference-centralization/preference-centralization.migration.ts

```diff
@@ -0,0 +1,154 @@
+/* eslint-disable no-console */
+import '../../src/config';
+
+import { NestFactory } from '@nestjs/core';
+import { SubscriberPreferenceRepository, NotificationTemplateRepository, PreferenceLevelEnum } from '@novu/dal';
+import {
+  UpsertPreferences,
+  UpsertWorkflowPreferencesCommand,
+  UpsertUserWorkflowPreferencesCommand,
+  UpsertSubscriberGlobalPreferencesCommand,
+  UpsertSubscriberWorkflowPreferencesCommand,
+} from '@novu/application-generic';
+import { buildWorkflowPreferencesFromPreferenceChannels, DEFAULT_WORKFLOW_PREFERENCES } from '@novu/shared';
+
+import { AppModule } from '../../src/app.module';
+
+const BATCH_SIZE = 100;
+
+const sleep = (ms: number) => {
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
+};
+
+/**
+ * Migration to centralize workflow and subscriber preferences.
+ * - subscriber global preference
+ *    -> preferences with subscriber global type
+ * - subscriber workflow preferences
+ *    -> preferences with subscriber workflow type
+ * - workflow preferences
+ *   -> preferences with workflow-resource type
+ *   -> preferences with user-workflow type
+ */
+export async function preferenceCentralization() {
```

Comment:
We need one more safeguard, that is the ability to stop, resume or rerun the script in case of unexpected failure, or timeout.

An easy way to do that is to keep the last cursor subscriberId or notificationTemplateId, print it and then allow the script to accept it as an argument so that the cursoring starts from there.

Having said that, cursors need to operate on sorted result sets.

Reply:
Great callout, I added a "last cursor", here are some sample results:

_SIGINT during workflow pref migration:_
```bash
> cross-env NODE_ENV=local MIGRATION=true ts-node --transpileOnly ./migrations/preference-centralization/preference-centralization.migration.ts
...
start migration - preference centralization
start workflow preference migration
^C
Gracefully shutting down from SIGINT (Ctrl+C)
Last processed workflow preference ID: 66ace0ae995ad7c319ef0f6b
```

_Completed migration, starting from last cursor:_
```bash
> npm run migration -- ./migrations/preference-centralization/preference-centralization.migration.ts 6737cd1d4b4d0b4f93e86575 6733b59bd496a30c6c01b6fa

> @novu/api@2.0.1 migration
> cross-env NODE_ENV=local MIGRATION=true ts-node --transpileOnly ./migrations/preference-centralization/preference-centralization.migration.ts 6737cd1d4b4d0b4f93e86575 6733b59bd496a30c6c01b6fa
...
start migration - preference centralization
start workflow preference migration
Starting from workflow preference ID: 6737cd1d4b4d0b4f93e86575
end workflow preference migration
{
  counter: {
    subscriberGlobal: { success: 0, error: 0 },
    subscriberWorkflow: { success: 0, error: 0 },
    subscriberUnknown: { success: 0, error: 0 },
    workflow: { success: 0, error: 0 }
  }
}
start subscriber preference migration
Starting from subscriber preference ID: 6733b59bd496a30c6c01b6fa
end subscriber preference migration
end migration - preference centralization
{
  counter: {
    subscriberGlobal: { success: 0, error: 0 },
    subscriberWorkflow: { success: 0, error: 0 },
    subscriberUnknown: { success: 0, error: 0 },
    workflow: { success: 0, error: 0 }
  }
}
Processed workflow preference with ID: undefined
Processed subscriber preference with ID: undefined
^C
Gracefully shutting down from SIGINT (Ctrl+C)
```

Reply:
Both of the sorted keys are indexed.